### PR TITLE
Add quantity selector component

### PR DESCRIPTION
**Pull request by vitoriaHeliane**
#### What is the purpose of this pull request?
Add the quantity selector component.

#### What problem is this solving?
Add the first version of the quantity selector component.

#### Screenshots or example usage
https://vitoria--storecomponents.myvtex.com/getting-started

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/theme.css [hunk 1]
```diff
@@ -10,3 +10,8 @@ html, body, #render-container, #render-provider, #render-provider > article {
 body {
   background-color: white;
 }
+
```
Comment (vitoriaHeliane) [inline react/theme.css]: Coloquei isso para as setinhas do input ficarem sempre ativas.

File: react/QtdSelector.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* The quantity selector component. It receives the maximous quantity that a client
```
Comment (brunojdo) [inline react/QtdSelector.js]: Remove "The" Quantity selector. Change `maximous` to `maximum`

Comment (brunojdo) [reply react/QtdSelector.js]: `the current quantity that the client already put in it and two functions, one to call when the limit is reached and another to call when the quantity is updated.` I think this shouldn't be in the class doc. Put on `handleChange` function

Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* The quantity selector component.
+* It constrols how many products a client wants to buy.
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: ⛏ "controls"

File: react/QtdSelector.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+*/
+class QtdSelector extends Component {
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: The component's name is a bit odd. "Qtd" is an abbreviation of the portuguese word "Quantidade". Consider using `QuantitySelector` instead.

File: react/QtdSelector.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  constructor(props) {
+    super(props)
+    this.handleChange = this.handleChange.bind(this)
+  }
+
+  handleChange(event) {
+    const qtdMax = this.props.qtdMax
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: Same comment about "qtd".

File: react/QtdSelector.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+* The quantity selector component.
+* It constrols how many products a client wants to buy.
+*/
+class QtdSelector extends Component {
+  constructor(props) {
+    super(props)
+    this.handleChange = this.handleChange.bind(this)
+  }
+
+  handleChange(event) {
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: Bind the function using an arrow binding here.
Ask @estacioneto about it.

File: react/QtdSelector.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    const qtdMax = this.props.qtdMax
+    var quantity = event.target.value
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: Use `let` instead of `var`.

File: react/QtdSelector.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      quantity = qtdMax
+      this.props.onMaxReached()
+    } else if (quantity < 0) {
+      quantity = 0
+    }
+    this.props.updateTotalPrice(quantity)
+  }
+
+  render() {
+    const {qtdMax, currentQtd} = this.props
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: Same comment about "qtd".

File: react/QtdSelector.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  }
+}
+
+QtdSelector.propTypes = {
+  /** The product's maximum quantity that the client can buy */
+  qtdMax: PropTypes.number.isRequired,
+  /** This is the current quantity to be setted as the initial value */
+  currentQtd: PropTypes.number,
+  /** This function is called when the client set the quantity selector. It is useful to set the
+  total price in the parent for exemple */
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: ⛏ "example"

File: react/QtdSelector.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  updateTotalPrice: PropTypes.func,
+  /** This function is useful to tell the user that can't buy more itens then the maximum limit */
```
Comment (leticiawanderley) [inline react/QtdSelector.js]: ⛏ "items"
⛏ "than"

Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 9 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 8 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 7 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 6 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 5 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 4 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QtdSelector.js (hunk 3 above)
Comment (vitoriaHeliane) [reply react/QtdSelector.js]: Done!

File: react/QuantitySelector.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+QuantitySelector.propTypes = {
+  /** The product's maximum quantity that the client can buy */
+  maxQuantity: PropTypes.number.isRequired,
+  /** This is the current quantity to be setted as the initial value */
+  currentQuantity: PropTypes.number,
+  /** This function is called when the client set the quantity selector. It is useful to set the
+  total price in the parent for example */
+  updateTotalPrice: PropTypes.func,
```
Comment (klzns) [inline react/QuantitySelector.js]: `isRequired`

File: react/QuantitySelector.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    return (<input className="f4 dark-gray br2-l" type="number" name="quantity"
+      min="0" max={maxQuantity} value={currentQuantity} onChange={this.handleChange} />)
+  }
+}
+
+QuantitySelector.propTypes = {
+  /** The product's maximum quantity that the client can buy */
+  maxQuantity: PropTypes.number.isRequired,
+  /** This is the current quantity to be setted as the initial value */
+  currentQuantity: PropTypes.number,
```
Comment (klzns) [inline react/QuantitySelector.js]: `isRequired`

File: react/QuantitySelector.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  /** This function is called when the client set the quantity selector. It is useful to set the
+  total price in the parent for example */
+  updateTotalPrice: PropTypes.func,
+  /** This function is useful to tell the user that can't buy more items than the maximum limit */
+  onMaxReached: PropTypes.func,
```
Comment (klzns) [inline react/QuantitySelector.js]: `isRequired`

File: react/QuantitySelector.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* The quantity selector component.
+* It controls how many products a client wants to buy.
+*/
+class QuantitySelector extends Component {
+  constructor(props) {
```
Comment (klzns) [inline react/QuantitySelector.js]: Retirar

Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js (hunk 13 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js (hunk 12 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js (hunk 11 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* The quantity selector component.
+*/
+class QuantitySelector extends Component {
+  constructor(props) {
```
Comment (klzns) [inline react/QuantitySelector.js]: Retirar

Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* The quantity selector component.
```
Comment (brunojdo) [inline react/QuantitySelector.js]: Quantity selector component

File: react/QuantitySelector.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+  render() {
+    const {maxQuantity, currentQuantity} = this.props
+    return (<input className="f4 dark-gray br2-l" type="number" name="quantity"
+      min="0" max={maxQuantity} value={currentQuantity} onChange={this.handleChange} />)
+  }
+}
+
+QuantitySelector.propTypes = {
+  /** The product's maximum quantity that the client can buy */
```
Comment (brunojdo) [inline react/QuantitySelector.js]: Product's maximum ... ⛏

Comment (brunojdo) [reply react/QuantitySelector.js]: Write just: 
"Current quantity to be setted as the initial value"
"Called when the client set the quantity selector"
"Define if can buy more items than the maximum limit"

File: react/QuantitySelector.js (hunk 16 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js (hunk 17 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

**Comment by ClaudivanFilho** [comment]
Tá quebrado Vitória 
screenshot from 2018-04-12 17-46-06

File: .gitignore [hunk 18]
```diff
@@ -34,3 +34,4 @@ npm-debug.log
 .build/
 lib
 .eslintrc
+package-lock.json
```
Comment (salesfelipe) [inline .gitignore]: Não acho que o package-lock.json deveria estar no gitignore.

File: react/QuantitySelector.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* Quantity selector component.
+*/
+class QuantitySelector extends Component {
+  handleChange = (event) => {
+    const maxQuantity = this.props.maxQuantity
```
Comment (salesfelipe) [inline react/QuantitySelector.js]: Melhor seguir esse padrão:

const { maxQuantity, updateTotalPrice } = this.props

File: react/QuantitySelector.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    let quantity = event.target.value
+    if (quantity > maxQuantity) {
+      quantity = maxQuantity
+      this.props.onMaxReached()
+    } else if (quantity < 0) {
+      quantity = 0
+    }
+    this.props.updateTotalPrice(quantity)
```
Comment (salesfelipe) [inline react/QuantitySelector.js]: Aproveita e muda esse pra:

updateTotalPrice(quantity)

File: react/QuantitySelector.js (hunk 19 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: .gitignore (hunk 18 above)
Comment (vitoriaHeliane) [reply .gitignore]: Isso foi a pedido de Estácio.

Comment (estacioneto) [reply .gitignore]: Pode tirar. Apesar de ficarem os arquivos `package-lock.json` e `yarn-lock.json` , eles ajudam

File: react/QuantitySelector.js [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+      quantity = 0
+    }
+    event.target.value = quantity
+    onQuantityChange(quantity)
+  }
+
+  render() {
+    const { maxQuantity, currentQuantity } = this.props
+    return (<input className="f4 dark-gray br2-l" type="number" name="quantity"
+      min="0" max={maxQuantity} value={currentQuantity} onChange={this.handleChange} />)
```
Comment (estacioneto) [inline react/QuantitySelector.js]: Se o input já tem min, precisa do min ali na função `handleChange` ?

File: react/first.js [hunk 22]
```diff
@@ -1,36 +1,40 @@
 import React from 'react'
```
Comment (estacioneto) [inline react/first.js]: Esse arquivo foi removido na master. Faz merge com master e coloca esses testes do componente no arquivo `index.js`

File: react/TableShoppingCart.js [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+            <td className="pv2 ph3">{name}</td>
+            <td className="pv2 ph3">R$ {sellingPrice}</td>
+            <td className="pv2 ph3">
+              <QuantitySelector
+                maxQuantity={maxQuantity}
+                currentQuantity={currentQuantity}
+                onQuantityChange={this.updateTotalPrice}
+                onMaxReached={this.onMaxReached} />
+            </td>
+            <td className="pv2 ph3">R$ {totalPrice}</td>
```
Comment (estacioneto) [inline react/TableShoppingCart.js]: Não necessariamente é `R$` . Vê como é feito no `Price.js` (usando `intl.formatNumber`)

Comment (estacioneto) [reply react/TableShoppingCart.js]: Muda nos outros locais tb :)

File: react/QuantitySelector.js [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+    const { maxQuantity, onQuantityChange, onMaxReached } = this.props
+    let value = event.target.value
+    let quantity = value ? parseInt(value, 10) : 0
+    if (quantity > maxQuantity) {
+      quantity = maxQuantity
+      onMaxReached()
+    } else if (quantity < 0) {
+      quantity = 0
+    }
+    event.target.value = quantity
```
Comment (lucasecdb) [inline react/QuantitySelector.js]: Avoid altering the DOM outside the `render` method, you can use the component state here to control the input value. Read more about it [here](https://reactjs.org/docs/forms.html), under the "Controlled Components" section

File: react/QuantitySelector.js [hunk 25]
```diff
@@ -15,7 +15,7 @@ class QuantitySelector extends Component {
     } else if (quantity < 0) {
       quantity = 0
     }
-    event.target.value = quantity
+    this.setState({ currentQuantity: quantity })
```
Comment (lucasecdb) [inline react/QuantitySelector.js]: You also need to use the value on `render`. Set the initial state with `getDerivedStateFromProps`, [check it's documentation](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops).

File: react/theme.css [hunk 26]
```diff
@@ -10,3 +10,8 @@ html, body, #render-container, #render-provider, #render-provider > article {
...
+input[type=number]::-webkit-inner-spin-button,
+input[type=number]::-webkit-outer-spin-button {
+   opacity: 1;
```
Comment (leticiawanderley) [inline react/theme.css]: The class `o-100` from `vtex-tachyons` has a similar behaviour.
https://github.com/vtex/vtex-tachyons/blob/master/tachyons.css#L1304

File: react/QuantitySelector.js [hunk 27]
```diff
@@ -0,0 +1,40 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+
+/**
+* Quantity selector component.
+*/
+class QuantitySelector extends Component {
+  handleChange = (event) => {
```
Comment (leticiawanderley) [inline react/QuantitySelector.js]: ⛏ no need to use parenthesis

File: react/QuantitySelector.js [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+    } else if (quantity < 0) {
+      quantity = 0
+    }
+    this.setState({ currentQuantity: quantity })
+    onQuantityChange(quantity)
+  }
+
+  render() {
+    const { maxQuantity, currentQuantity } = this.props
+    return (<input className="f4 dark-gray br2-l" type="number" name="quantity"
```
Comment (leticiawanderley) [inline react/QuantitySelector.js]: ⛏ not needed parenthesis

File: react/QuantitySelector.js [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+    const { maxQuantity, onQuantityChange, onMaxReached } = this.props
+    let value = event.target.value
+    let quantity = value ? parseInt(value, 10) : 0
+    if (quantity > maxQuantity) {
+      quantity = maxQuantity
+      onMaxReached()
+    } else if (quantity < 0) {
+      quantity = 0
+    }
+    this.setState({ currentQuantity: quantity })
```
Comment (leticiawanderley) [inline react/QuantitySelector.js]: Is this `currentQuantity` state variable ever used?
The `render` method uses the one that is inside the `props`.

File: react/QuantitySelector.js (hunk 28 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/QuantitySelector.js (hunk 27 above)
Comment (vitoriaHeliane) [reply react/QuantitySelector.js]: Done!

File: react/theme.css (hunk 26 above)
Comment (vitoriaHeliane) [reply react/theme.css]: Done!

**Pull request merged**